core/src/banking_stage.rs

```diff
@@ -803,11 +887,32 @@ impl BankingStage {
     }
 }
 
+#[cfg_attr(feature = "dev-context-only-utils", qualifiers(pub))]
+pub(crate) fn update_bank_forks_and_poh_recorder_for_new_tpu_bank(
+    bank_forks: &RwLock<BankForks>,
+    poh_recorder: &RwLock<PohRecorder>,
+    tpu_bank: Bank,
+    track_transaction_indexes: bool,
+) {
+    // A write lock for the poh recorder must be grabbed for the entire duration of inserting new
+    // tpu bank into the bank forks. That's because any buffered transactions could immediately be
+    // executed after the bank forks update, when unified scheduler is enabled for block
+    // production. If transactions were executed prematurely, the unified scheduler would be hit
+    // with false errors due to having no bank in the poh recorder otherwise.
+    let mut poh_recorder = poh_recorder.write().unwrap();
+
+    let tpu_bank = bank_forks
+        .write()
+        .unwrap()
+        .insert_with_scheduling_mode(SchedulingMode::BlockProduction, tpu_bank);
+    poh_recorder.set_bank(tpu_bank, track_transaction_indexes);
```

Comment:
this

Reply:
a concern here is that elsewhere we grab these locks in the opposite order.
I'm fairly certain we don't but given these are both so extensively used (smelly) I'm not convinced entirely that we do not.

Eventually there should be some other delivery system for TPU banks, other than locking poh. I'm not quite sure what that should look like though. Any thoughts?

Reply:
> a concern here is that elsewhere we grab these locks in the opposite order.

how about this? this removes the opposite locking order: d804e35bd7082696ce7c9356ce628fde420e5756 (this builds on top of revert: dbf4db47cff31856b8ccd8049d1f5f3cd5c60458)

> Eventually there should be some other delivery system for TPU banks, other than locking poh. I'm not quite sure what that should look like though. Any thoughts?

hmm, i don't think locking poh_recorder is so bad in general.. come to think of it, it's a bit odd that poh is being a singleton. i think we just can instantiate them on demand for new tpu bank (this most of time will end up there's one instance effectively given any time of leader schedule. but this should simplify or relax locking requirements).